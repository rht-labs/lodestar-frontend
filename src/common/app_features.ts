// All roles defined by the backend to be used in the frontend should be enumerated in this object.

export const APP_FEATURES = {
  admin: 'admin',
  writer: 'writer',
  reader: 'reader',
  notifications: 'notifications',
  engagementCardIcons: 'engagementCardIcons',
  resetUser: 'resetUser',
};

export type AppFeature = keyof typeof APP_FEATURES;
