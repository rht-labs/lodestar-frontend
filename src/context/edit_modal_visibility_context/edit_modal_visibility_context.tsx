import React, { useState, useCallback } from 'react';

export interface IModalVisibilityContext {
  activeModalKey: string;
  requestOpen: (requestingModalKey: string) => void;
  requestClose: () => void;
}

export const ModalVisibilityContext = React.createContext<
  IModalVisibilityContext
>({
  activeModalKey: null,
  requestOpen: () => null,
  requestClose: () => null,
});

const { Provider } = ModalVisibilityContext;

export const ModalVisibilityProvider = ({ children }) => {
  const [activeModalKey, setActiveModalKey] = useState<string>(null);
  const requestOpen = useCallback(
    (requestingModalKey: string) => {
      setActiveModalKey(requestingModalKey);
    },
    [setActiveModalKey]
  );

  const requestClose = useCallback(() => setActiveModalKey(null), [
    setActiveModalKey,
  ]);

  return (
    <Provider
      value={{
        requestClose,
        activeModalKey,
        requestOpen,
      }}
    >
      {children}
    </Provider>
  );
};
