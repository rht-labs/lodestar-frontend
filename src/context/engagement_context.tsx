import React, { createContext, useEffect, useState, useCallback } from 'react';
import { Engagement } from '../models/engagement';
import { EngagementRepository } from '../repositories/engagement/engagement_repository';

export interface EngagementContext {
  getEngagements: () => void;
  engagements: Engagement[];
}

export const EngagementContext = createContext<EngagementContext>({
  getEngagements: async () => [],
  engagements: [],
});
const { Provider } = EngagementContext;

export const EngagementProvider = ({
  children,
  engagementRepository,
}: {
  children: React.ReactChild;
  engagementRepository: EngagementRepository;
}) => {
  const [engagements, setEngagements] = useState<Engagement[]>([]);

  const fetchEngagements = useCallback(async () => {
    const engagements = await engagementRepository.fetchEngagements();
    setEngagements(engagements);
  }, [engagementRepository]);

  useEffect(() => {
    fetchEngagements();
  }, [fetchEngagements]);

  return (
    <Provider
      value={{
        engagements,
        getEngagements: fetchEngagements,
      }}
    >
      {children}
    </Provider>
  );
};
