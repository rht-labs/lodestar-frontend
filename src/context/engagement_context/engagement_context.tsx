import React, { createContext, useEffect, useRef } from 'react';
import { Engagement } from '../../schemas/engagement';
import { useState, useCallback, useReducer } from 'react';
import {
  engagementFormReducer,
  getInitialState,
} from './engagement_form_reducer';
import { useFeedback, AlertType } from '../feedback_context/feedback_context';
import { EngagementFormConfig } from '../../schemas/engagement_config';
import { AlreadyExistsError } from '../../services/engagement_service/engagement_service_errors';
import { Logger } from '../../utilities/logger';
import { AuthenticationError } from '../../services/auth_service/auth_errors';
import { useSession } from '../auth_context/auth_context';
import {
  EngagementPoll,
  EngagementPollIntervalStrategy,
} from '../../schemas/engagement_poll';
import { EngagementService } from '../../services/engagement_service/engagement_service';

export interface EngagementContext {
  getEngagements: () => Promise<Engagement[]>;
  currentEngagementChanges?: Engagement;
  currentEngagement?: Engagement;
  setCurrentEngagement: (Engagement: Engagement) => void;
  engagements?: Engagement[];
  requiredFields: string[];
  getEngagement: (
    customerName: string,
    projectName: string
  ) => Promise<Engagement>;
  getConfig: () => void;
  createEngagement: (data: any) => Promise<Engagement>;
  saveEngagement: (data: any) => Promise<void>;
  updateEngagementFormField: (fieldName: string, payload: any) => void;
  missingRequiredFields: string[];
  isLaunchable: boolean;
  formOptions?: EngagementFormConfig;
  error: any;
  isLoading: boolean;
  launchEngagement: (data: any) => Promise<void>;
  createEngagementPoll: (engagement: Engagement) => Promise<EngagementPoll>;
}

export type CreateEngagementParams = Pick<
  Engagement,
  'project_name' | 'customer_name' | 'engagement_region'
>;

const requiredFields = [
  'customer_contact_email',
  'customer_contact_name',
  'customer_name',
  'end_date',
  'start_date',
  'engagement_lead_email',
  'technical_lead_email',
  'engagement_lead_name',
  'technical_lead_name',
  'ocp_cloud_provider_name',
  'ocp_cloud_provider_region',
  'ocp_version',
  'ocp_cluster_size',
  'ocp_persistent_storage_size',
  'ocp_sub_domain',
  'project_name',
];
export const EngagementContext = createContext<EngagementContext>(null);

const { Provider } = EngagementContext;

export const EngagementProvider = ({
  children,
  engagementService,
}: {
  children: React.ReactChild;
  engagementService: EngagementService;
}) => {
  const feedbackContext = useFeedback();
  const [formOptions, setFormOptions] = useState<EngagementFormConfig>();

  const [error] = useState<any>();
  const [isLoading] = useState<boolean>(false);
  const [engagements, setEngagements] = useState<Engagement[]>([]);
  const [currentEngagement, setCurrentEngagement] = useState<
    Engagement | undefined
  >();
  const [currentEngagementChanges, dispatch] = useReducer<
    (state: any, action: any) => any
  >(engagementFormReducer(formOptions), engagementFormReducer(formOptions)());

  const authContext = useSession();

  const _handleErrors = useCallback(
    async error => {
      Logger.instance.debug('EngagementContext:_handleErrors', error);
      if (error instanceof AuthenticationError) {
        if (!(await authContext.checkIsAuthenticated())) {
          authContext.setAuthError(error);
        }
      } else {
        throw error;
      }
    },
    [authContext]
  );

  useEffect(() => {
    _validateAuthStatusRef.current = async () => {
      if (!(await authContext.checkIsAuthenticated())) {
        throw new AuthenticationError();
      }
    };
  }, [authContext]);

  const _validateAuthStatusRef = useRef(async () => {});

  const getConfig = useCallback(async () => {
    await _validateAuthStatusRef.current();
    const data = await engagementService.getConfig();
    setFormOptions(data);
  }, [engagementService]);

  useEffect(() => {
    dispatch({
      type: 'switch_engagement',
      payload: getInitialState(currentEngagement),
    });
  }, [currentEngagement, formOptions]);

  const _updateEngagementInPlace = useCallback(
    engagement => {
      const oldEngagementIndex = engagements.findIndex(comparisonEngagement => {
        if (
          engagement?.project_name &&
          engagement?.customer_name &&
          engagement?.project_name === comparisonEngagement?.project_name &&
          engagement?.customer_name === comparisonEngagement?.customer_name
        ) {
          return true;
        }
        return false;
      });
      const oldEngagement = engagements[oldEngagementIndex];
      if (oldEngagementIndex > -1) {
        const newEngagements = [...engagements];
        newEngagements.splice(oldEngagementIndex, 1, engagement);
        setEngagements(newEngagements);
      }
      return oldEngagement;
    },
    [engagements]
  );

  const fetchEngagements = useCallback(async () => {
    try {
      await _validateAuthStatusRef.current();
      feedbackContext.showLoader();
      const engagements = await engagementService.fetchEngagements();
      setEngagements(engagements);
      feedbackContext.hideLoader();
      return engagements;
    } catch (e) {
      feedbackContext.showAlert(
        'Something went wrong while getting the engagements',
        AlertType.error,
        true
      );
      feedbackContext.hideLoader();
      await _handleErrors(e);
    }
  }, [engagementService, feedbackContext, _handleErrors]);

  const _refreshEngagementData = useCallback(
    async (engagement: Engagement) => {
      await _validateAuthStatusRef.current();
      try {
        const refreshedEngagement = await engagementService.getEngagementByCustomerAndProjectName(
          engagement?.customer_name,
          engagement?.project_name
        );
        _updateEngagementInPlace(refreshedEngagement);
        setCurrentEngagement(refreshedEngagement);
      } catch (e) {
        _handleErrors(e);
      }
    },
    [_updateEngagementInPlace, engagementService, _handleErrors]
  );

  const _checkHasUpdateRef = useRef(async () => false);

  useEffect(() => {
    const checkUpdate = async () => {
      return await engagementService.checkHasUpdates(currentEngagement);
    };
    _checkHasUpdateRef.current = checkUpdate;
  }, [currentEngagement, engagementService]);

  const createEngagementPoll = useCallback(
    async (engagement: Engagement): Promise<EngagementPoll> => {
      await _validateAuthStatusRef.current();
      return new EngagementPoll(
        new EngagementPollIntervalStrategy(
          setInterval(async () => {
            await _validateAuthStatusRef.current();
            const hasUpdates = await _checkHasUpdateRef.current();
            if (hasUpdates) {
              feedbackContext.showAlert(
                'Another user edited this engagement. In order to continue, you must refresh the page. By refreshing, your unsaved changes will be overwritten."',
                AlertType.error,
                false,
                [
                  {
                    title: 'Refresh',
                    action: () => _refreshEngagementData(engagement),
                  },
                ]
              );
            }
          }, 5000)
        )
      );
    },
    [_refreshEngagementData, feedbackContext]
  );

  const getEngagement = useCallback(
    async (customerName: string, projectName: string) => {
      try {
        let availableEngagements = engagements ?? [];
        let cachedEngagement = availableEngagements?.find(
          engagement =>
            engagement?.customer_name === customerName &&
            engagement?.project_name === projectName
        );
        if (cachedEngagement !== null) {
          setCurrentEngagement(cachedEngagement);
        }
        let fetchedEngagement = await engagementService.getEngagementByCustomerAndProjectName(
          customerName,
          projectName
        );
        setCurrentEngagement(fetchedEngagement);
        return fetchedEngagement;
      } catch (e) {
        try {
          await _handleErrors(e);
        } catch (e) {
          feedbackContext.showAlert(
            'There was a problem fetching this engagement',
            AlertType.error
          );
        }
      }
    },
    [engagements, feedbackContext, _handleErrors, engagementService]
  );

  const _addNewEngagement = useCallback(
    async (newEngagement: Engagement) => {
      try {
        const newEngagementList = [newEngagement, ...(engagements ?? [])];
        setEngagements(newEngagementList);
      } catch (e) {
        await _handleErrors(e);
        Logger.instance.error(e);
      }
    },
    [engagements, _handleErrors]
  );

  const createEngagement = useCallback(
    async (data: CreateEngagementParams) => {
      feedbackContext.showLoader();
      try {
        await _validateAuthStatusRef.current();
        const engagement = await engagementService.createEngagement(data);
        _addNewEngagement(engagement);
        setEngagements([...(engagements ?? []), engagement]);
        feedbackContext.hideLoader();
        feedbackContext.showAlert(
          'Your engagement has been successfully created',
          AlertType.success
        );
        setCurrentEngagement(engagement);
        return engagement;
      } catch (e) {
        feedbackContext.hideLoader();
        let errorMessage;
        if (e instanceof AlreadyExistsError) {
          errorMessage =
            'This client already has a project with that name. Please choose a different project name.';
        } else {
          try {
            await _handleErrors(e);
          } catch (e) {
            errorMessage =
              'There was an issue with creating your engagement. Please follow up with an administrator if this continues.';
          }
        }

        feedbackContext.showAlert(errorMessage, AlertType.error);
        await _handleErrors(e);
      }
    },
    [
      engagementService,
      _addNewEngagement,
      feedbackContext,
      engagements,
      _handleErrors,
    ]
  );

  const _checkLaunchReady = useCallback(() => {
    let result = requiredFields.every(
      o =>
        typeof currentEngagementChanges[o] === 'boolean' ||
        typeof currentEngagementChanges[o] === 'number' ||
        !!currentEngagementChanges[o]
    );
    return result;
  }, [currentEngagementChanges]);

  const missingRequiredFields = useCallback(() => {
    return requiredFields.filter(
      field =>
        currentEngagement?.[field] !== 'boolean' &&
        currentEngagement?.[field] !== 'number' &&
        !currentEngagement?.[field]
    );
  }, [currentEngagement]);

  const saveEngagement = useCallback(
    async (data: Engagement) => {
      feedbackContext.showLoader();
      const oldEngagement = _updateEngagementInPlace(data);
      try {
        await _validateAuthStatusRef.current();
        const returnedEngagement = await engagementService.saveEngagement(data);
        feedbackContext.showAlert(
          'Your updates have been successfully saved.',
          AlertType.success
        );
        feedbackContext.hideLoader();
        _updateEngagementInPlace(returnedEngagement);
        setCurrentEngagement(returnedEngagement);
      } catch (e) {
        _updateEngagementInPlace(oldEngagement);
        feedbackContext.hideLoader();
        let errorMessage =
          'There was an issue with saving your changes. Please followup with an administrator if this continues.';
        if (e instanceof AlreadyExistsError) {
          // If there is no mongo id associated with the engagement, then it is being committed to the backend for the first time. It is a net new engagement.
          if (!!data.mongo_id) {
            errorMessage =
              'The path that you input is already taken.  Please update and try saving again.';

            // Otherwise, we are saving an existing engagement.
          } else {
            await _handleErrors(e);
            errorMessage =
              'Your changes could not be saved. Another user has modified this data. Please refresh your data in order to make changes.';
          }
        }
        feedbackContext.showAlert(errorMessage, AlertType.error);
      }
    },
    [
      engagementService,
      _updateEngagementInPlace,
      feedbackContext,
      _handleErrors,
    ]
  );

  const updateEngagementFormField = useCallback(
    (fieldName: string, value: any) => {
      dispatch({ type: fieldName, payload: value });
    },
    [dispatch]
  );

  const launchEngagement = useCallback(
    async (data: any) => {
      if (!_checkLaunchReady()) {
        throw Error(
          'This engagement does not have the required fields to launch'
        );
      }
      feedbackContext.showLoader();
      const oldEngagement = _updateEngagementInPlace(data);
      try {
        await _validateAuthStatusRef.current();
        const returnedEngagement = await engagementService.launchEngagement(
          data
        );
        _updateEngagementInPlace(returnedEngagement);
        setCurrentEngagement(returnedEngagement);
        feedbackContext.hideLoader();
        feedbackContext.showAlert(
          'You have successfully launched your engagement!',
          AlertType.success
        );
      } catch (e) {
        _updateEngagementInPlace(oldEngagement);
        feedbackContext.hideLoader();
        feedbackContext.showAlert(
          'We were unable to launch your engagement. Please follow up with an administrator if this continues.',
          AlertType.error,
          false
        );
        await _handleErrors(e);
      }
    },
    [
      _updateEngagementInPlace,
      _checkLaunchReady,
      engagementService,
      feedbackContext,
      _handleErrors,
    ]
  );
  return (
    <Provider
      value={{
        createEngagementPoll,
        requiredFields,
        currentEngagement,
        missingRequiredFields: missingRequiredFields(),
        getConfig,
        isLaunchable: _checkLaunchReady(),
        setCurrentEngagement,
        engagements,
        getEngagement,
        error,
        currentEngagementChanges: {
          ...currentEngagement,
          ...currentEngagementChanges,
        },
        formOptions,
        isLoading,
        updateEngagementFormField,
        getEngagements: fetchEngagements,
        createEngagement,
        saveEngagement,
        launchEngagement,
      }}
    >
      {children}
    </Provider>
  );
};

// EngagementProvider.whyDidYouRender = false;
