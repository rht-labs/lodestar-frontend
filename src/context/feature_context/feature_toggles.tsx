import React, { useCallback, useEffect, useState } from 'react';
import { useLocation } from 'react-router';
import {
  getFeaturesFromVersion,
  FEATURE_VERSION_MAP,
} from '../../common/version_feature_factory';
import { IAuthContext } from '../auth_context/auth_context';
import { IVersionContext } from '../version_context/version_context';
import qs from 'query-string';
import { Config } from '../../schemas/config';

export interface IFeatureToggleContext {
  features: string[];
  isLoading: boolean;
  hasFeature: (name: string) => boolean;
}

export const FeatureToggleContext = React.createContext<IFeatureToggleContext>({
  features: [],
  hasFeature: () => false,
  isLoading: false,
});

export interface IFeatureToggleProvider {
  features?: string[];
  authContext: IAuthContext;
  versionContext: IVersionContext;
  config?: Config;
  children: any;
}

export const FeatureToggles = ({
  children,
  features,
  authContext,
  versionContext,
  config,
}: IFeatureToggleProvider) => {
  let version = versionContext?.versions?.mainVersion?.value;

  const location = useLocation();
  if (config?.allowVersionOverride) {
    const query = qs.parse(location.search);
    const queryVersion = query['lodestar-version'];
    if (queryVersion) {
      version = Array.isArray(queryVersion) ? queryVersion[0] : queryVersion;
    }
  }
  const versionFeatures = getFeaturesFromVersion(version, FEATURE_VERSION_MAP);
  const roleFeatures = (authContext?.sessionData?.roles ?? []).concat(
    features ?? []
  );
  const [hasFetchedVersions, setHasFetchedVersions] = useState(false);
  useEffect(() => {
    if (!hasFetchedVersions) {
      versionContext?.fetchVersions?.();
      setHasFetchedVersions(true);
    }
  }, [versionContext, hasFetchedVersions]);
  const allFeatures = React.useMemo(
    () => [...versionFeatures, ...roleFeatures],
    [versionFeatures, roleFeatures]
  );
  const hasFeature = useCallback(
    (name: string) => {
      return name && allFeatures.includes(name);
    },
    [allFeatures]
  );
  const isLoading = !versionContext?.versions;

  return (
    <FeatureToggleContext.Provider
      value={{ features: allFeatures, hasFeature, isLoading }}
    >
      {children}
    </FeatureToggleContext.Provider>
  );
};
