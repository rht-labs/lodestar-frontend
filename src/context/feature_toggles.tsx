import React, { useContext } from 'react';
import { SessionContext } from './session_context';

interface FeatureToggleContext {
  features: string[];
}

export const FeatureToggleContext = React.createContext<FeatureToggleContext>({
  features: [],
});

export const FeatureToggles = ({ children }: { children: React.ReactNode }) => {
  const sessionContext = useContext(SessionContext);
  const features = sessionContext.sessionData?.roles ?? [];
  return (
    <FeatureToggleContext.Provider value={{ features }}>
      {children}
    </FeatureToggleContext.Provider>
  );
};
