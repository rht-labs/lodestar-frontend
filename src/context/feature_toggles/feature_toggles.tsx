import React, { useCallback } from 'react';
import { useSession } from '../auth_context/auth_context';

interface FeatureToggleContext {
  features: string[];
  hasFeature: (name: string) => boolean;
}

export const FeatureToggleContext = React.createContext<FeatureToggleContext>({
  features: [],
  hasFeature: () => false,
});

export const FeatureToggles = ({
  children,
  features,
}: {
  children: React.ReactNode;
  features?: string[];
}) => {
  const authContext = useSession();
  const providedFeatures = features ?? authContext.sessionData?.roles ?? [];
  const hasFeature = useCallback(
    (name: string) => {
      return name && providedFeatures.includes(name);
    },
    [providedFeatures]
  );

  return (
    <FeatureToggleContext.Provider
      value={{ features: providedFeatures, hasFeature }}
    >
      {children}
    </FeatureToggleContext.Provider>
  );
};
