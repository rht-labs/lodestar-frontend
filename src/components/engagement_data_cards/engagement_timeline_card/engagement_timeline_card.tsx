import React, { useState } from 'react';
import { uuid } from 'uuidv4';
import { DataCard } from '../data_card';
import {Artifact, ArtifactType} from '../../../schemas/engagement';
import { EditButton } from '../../data_card_edit_button/data_card_edit_button';
import { useModalVisibility } from '../../../context/edit_modal_visibility_context/edit_modal_visibility_hook';
import {
  Table,
  TableVariant,
  TableHeader,
  TableBody,
  cellWidth,
} from '@patternfly/react-table';
import {
  Dropdown,
  KebabToggle,
  DropdownItem,
  EmptyState,
  EmptyStateIcon,
  Title,
  EmptyStateBody,
} from '@patternfly/react-core';
import { ArtifactEditModal } from '../../engagement_edit_modals/add_artifact_modal';
import { ClipboardCheckIcon } from '@patternfly/react-icons';
import { Feature } from '../../feature/feature';
import { useEngagements } from '../../../context/engagement_context/engagement_hook';
import { useEngagementArtifacts } from '../../../context/engagement_context/engagement_context';

export interface EngagementTimelineCardProps {
  artifacts: Artifact[];
  onClear: () => void;
  onSave: (artifacts: Array<Artifact>) => void;
}

const ARTIFACT_CRUD_MODAL = 'artifact_crud_modal';

export function EngagementTimelineCard() {
  const {
    currentChanges,
    saveEngagement,
    clearCurrentChanges,
  } = useEngagements();
  const { addArtifact, artifacts } = useEngagementArtifacts();
  const { requestOpen, activeModalKey, requestClose } = useModalVisibility();
  const [currentArtifact, setCurrentArtifact] = useState<Artifact>();

  const onEditArtifact = (artifact: Artifact) => {
    setCurrentArtifact(artifact);
    requestOpen(ARTIFACT_CRUD_MODAL);
  };

  const openArtifactModal = () => {
    requestOpen(ARTIFACT_CRUD_MODAL);
    setCurrentArtifact({ id: uuid() } as Artifact);
  };

  const onSave = (artifacts: Artifact[]) =>
    saveEngagement({ ...currentChanges, artifacts });

  const onFinishArtifactEdit = (artifact: Artifact) => {
    onSave(addArtifact(artifact));
  };

  return (
    <div>
      <ArtifactEditModal
        artifact={currentArtifact}
        isOpen={activeModalKey === ARTIFACT_CRUD_MODAL}
        onClose={() => {
          requestClose();
          clearCurrentChanges();
        }}
        onSave={onFinishArtifactEdit}
      />
      <DataCard
        title="Engagement Artifacts"
        trailingIcon={() => <div />}
        actionButton={() => (
          <EditButton
            text="Add an Artifact"
            onClick={openArtifactModal}
            data-testid="add-artifact-button"
          />
        )}
      >
        <EngagementTimelineCardBody
          artifacts={artifacts}
          onClear={clearCurrentChanges}
          onSave={artifacts => saveEngagement({ ...currentChanges, artifacts })}
          onAdd={openArtifactModal}
          editArtifact={onEditArtifact}
        />
      </DataCard>
    </div>
  );
}

function EngagementTimelineCardBody(
  props: EngagementTimelineCardProps & {
    editArtifact(artifact: Artifact): void;
    onAdd(): void;
  }
) {
  const [currentOpenDropdown, setCurrentOpenDropdown] = useState<number>();
  const columns = [
    { title: 'Type', transforms: [cellWidth(10)] },
    { title: 'Title', transforms: [cellWidth(20)] },
    { title: 'Description', transforms: [cellWidth('max')] },
    { title: 'Actions' },
  ];
  function getAbsoluteUrl(url: string): string {
    if (url?.includes?.('://')) {
      return url;
    } else {
      return `//${url}`;
    }
  }
  const actionItems = [
    <DropdownItem key="edit">
      <span data-testid="artifact-edit-button">Edit</span>
    </DropdownItem>,
  ];
  const rows =
    props?.artifacts?.map?.((artifact, idx) => [
      ArtifactType[artifact.type],
      {
        title: (
          <a
            target="_blank"
            rel="noopener noreferrer"
            href={getAbsoluteUrl(artifact.linkAddress)}
          >
            {artifact.title}
          </a>
        ),
      },
      artifact.description,
      {
        title: (
          <Feature name={'writer'}>
            <Dropdown
              isPlain
              dropdownItems={actionItems}
              isOpen={idx === currentOpenDropdown}
              onSelect={() => {
                setCurrentOpenDropdown(undefined);
                props.editArtifact(artifact);
              }}
              toggle={
                <KebabToggle
                  onToggle={() =>
                    currentOpenDropdown === idx
                      ? setCurrentOpenDropdown(undefined)
                      : setCurrentOpenDropdown(idx)
                  }
                  id={`toggle-id-${idx}`}
                  data-testid="artifact-action-kebab"
                />
              }
            />
          </Feature>
        ),
      },
    ]) ?? [];

  return props?.artifacts?.length > 0 ? (
    <Table
      aria-label="Engagement Artifacts"
      variant={TableVariant.compact}
      cells={columns}
      rows={rows}
    >
      <TableHeader />
      <TableBody />
    </Table>
  ) : (
    <EmptyState>
      <EmptyStateIcon
        icon={ClipboardCheckIcon}
        style={{ fontSize: '34px', margin: '0' }}
      />
      <Title headingLevel="h5" size="md" style={{ marginTop: '0' }}>
        No Artifacts Added
      </Title>
      <EmptyStateBody>
        <p>Click 'Add Artifact' button to start adding artifacts</p>
      </EmptyStateBody>
    </EmptyState>
  );
}
