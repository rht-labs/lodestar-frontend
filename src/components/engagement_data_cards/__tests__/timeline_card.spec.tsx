import React from 'react';
import {
  render,
  act,
  fireEvent,
  waitForDomChange,
} from '@testing-library/react';

import { EngagementTimelineCard } from '../engagement_timeline_card/engagement_timeline_card';
import { Artifact } from '../../../schemas/engagement';
import { ModalVisibilityContext } from '../../../context/edit_modal_visibility_context/edit_modal_visibility_context';
import { FeatureToggleContext } from '../../../context/feature_context/feature_toggles';
import { APP_FEATURES } from '../../../common/app_features';
import { EngagementContext } from '../../../context/engagement_context/engagement_context';

describe('Engagement Artifact Card', () => {
  test('matches snapshot', () => {
    const timelineCard = render(
      <EngagementContext.Provider
        value={{
          currentEngagement: {
            artifacts: new Array(20)
              .fill(null)
              .map(() => Artifact.fromFake(true)),
          },
        }}
      >
        <EngagementTimelineCard />
      </EngagementContext.Provider>
    );
    expect(timelineCard).toMatchSnapshot();
  });
  test('clicking add an artifact opens the edit artifact modal', async () => {
    const requestOpen = jest.fn();
    const card = render(
      <FeatureToggleContext.Provider
        value={{
          hasFeature: () => true,
          features: [APP_FEATURES.writer, APP_FEATURES.reader],
        }}
      >
        <ModalVisibilityContext.Provider
          value={{ requestOpen, activeModalKey: '', requestClose: () => {} }}
        >
          <EngagementTimelineCard></EngagementTimelineCard>
        </ModalVisibilityContext.Provider>
      </FeatureToggleContext.Provider>
    );
    act(() => {
      fireEvent.click(card.getByTestId('add-artifact-button'));
      expect(requestOpen).toHaveBeenCalled();
    });
  });
  test('clicking the edit artifact dropdown item opens the edit artifact modal', async () => {
    const requestOpen = jest.fn();
    const card = render(
      <FeatureToggleContext.Provider
        value={{
          hasFeature: () => true,
          features: [APP_FEATURES.writer, APP_FEATURES.reader],
        }}
      >
        <ModalVisibilityContext.Provider
          value={{ requestOpen, activeModalKey: '', requestClose: () => {} }}
        >
          <EngagementContext.Provider
            value={{
              currentEngagement: { artifacts: [Artifact.fromFake(true)] },
            }}
          >
            <EngagementTimelineCard></EngagementTimelineCard>
          </EngagementContext.Provider>
        </ModalVisibilityContext.Provider>
      </FeatureToggleContext.Provider>
    );
    await act(async () => {
      fireEvent.click(card.getByTestId('artifact-action-kebab'));
      await waitForDomChange({ container: card.container });
      fireEvent.click(card.getByTestId('artifact-edit-button'));
      expect(requestOpen).toHaveBeenCalled();
    });
  });
});
