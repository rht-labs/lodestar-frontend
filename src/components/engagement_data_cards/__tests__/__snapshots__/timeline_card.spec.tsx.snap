// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Engagement Artifact Card matches snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div>
        <article
          class="pf-c-card pf-m-compact"
          data-ouia-component-id="0"
          data-ouia-component-type="PF4/Card"
          data-ouia-safe="true"
          style="padding: 1rem;"
        >
          <div
            class="pf-c-card__title"
          >
            <div
              class="pf-l-level"
            >
              <div>
                <h3
                  class="pf-c-title pf-m-lg"
                  style="font-weight: normal;"
                >
                  Engagement Artifacts
                    
                  <div />
                </h3>
              </div>
              <div />
            </div>
          </div>
          <div
            class="pf-c-card__body"
          >
            <div
              class="pf-c-empty-state"
            >
              <div
                class="pf-c-empty-state__content"
              >
                <svg
                  aria-hidden="true"
                  class="pf-c-empty-state__icon"
                  fill="currentColor"
                  height="1em"
                  role="img"
                  style="font-size: 34px; margin: 0px;"
                  viewBox="0 0 384 512"
                  width="1em"
                >
                  <path
                    d="M336 64h-80c0-35.3-28.7-64-64-64s-64 28.7-64 64H48C21.5 64 0 85.5 0 112v352c0 26.5 21.5 48 48 48h288c26.5 0 48-21.5 48-48V112c0-26.5-21.5-48-48-48zM192 40c13.3 0 24 10.7 24 24s-10.7 24-24 24-24-10.7-24-24 10.7-24 24-24zm121.2 231.8l-143 141.8c-4.7 4.7-12.3 4.6-17-.1l-82.6-83.3c-4.7-4.7-4.6-12.3.1-17L99.1 285c4.7-4.7 12.3-4.6 17 .1l46 46.4 106-105.2c4.7-4.7 12.3-4.6 17 .1l28.2 28.4c4.7 4.8 4.6 12.3-.1 17z"
                    transform=""
                  />
                </svg>
                <h5
                  class="pf-c-title pf-m-md"
                  style="margin-top: 0px;"
                >
                  No Artifacts Added
                </h5>
                <div
                  class="pf-c-empty-state__body"
                >
                  <p>
                    Click 'Add Artifact' button to start adding artifacts
                  </p>
                </div>
              </div>
            </div>
          </div>
        </article>
      </div>
    </div>
  </body>,
  "container": <div>
    <div>
      <article
        class="pf-c-card pf-m-compact"
        data-ouia-component-id="0"
        data-ouia-component-type="PF4/Card"
        data-ouia-safe="true"
        style="padding: 1rem;"
      >
        <div
          class="pf-c-card__title"
        >
          <div
            class="pf-l-level"
          >
            <div>
              <h3
                class="pf-c-title pf-m-lg"
                style="font-weight: normal;"
              >
                Engagement Artifacts
                  
                <div />
              </h3>
            </div>
            <div />
          </div>
        </div>
        <div
          class="pf-c-card__body"
        >
          <div
            class="pf-c-empty-state"
          >
            <div
              class="pf-c-empty-state__content"
            >
              <svg
                aria-hidden="true"
                class="pf-c-empty-state__icon"
                fill="currentColor"
                height="1em"
                role="img"
                style="font-size: 34px; margin: 0px;"
                viewBox="0 0 384 512"
                width="1em"
              >
                <path
                  d="M336 64h-80c0-35.3-28.7-64-64-64s-64 28.7-64 64H48C21.5 64 0 85.5 0 112v352c0 26.5 21.5 48 48 48h288c26.5 0 48-21.5 48-48V112c0-26.5-21.5-48-48-48zM192 40c13.3 0 24 10.7 24 24s-10.7 24-24 24-24-10.7-24-24 10.7-24 24-24zm121.2 231.8l-143 141.8c-4.7 4.7-12.3 4.6-17-.1l-82.6-83.3c-4.7-4.7-4.6-12.3.1-17L99.1 285c4.7-4.7 12.3-4.6 17 .1l46 46.4 106-105.2c4.7-4.7 12.3-4.6 17 .1l28.2 28.4c4.7 4.8 4.6 12.3-.1 17z"
                  transform=""
                />
              </svg>
              <h5
                class="pf-c-title pf-m-md"
                style="margin-top: 0px;"
              >
                No Artifacts Added
              </h5>
              <div
                class="pf-c-empty-state__body"
              >
                <p>
                  Click 'Add Artifact' button to start adding artifacts
                </p>
              </div>
            </div>
          </div>
        </div>
      </article>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
